### PR TITLE
Introduce new PluginTemplateSettingPanel slot

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
This PR introduces a new slot in the Template Sidebar below the main information like the Template Card.

Should we add this slot in `block-editor` package in order to be reused when we edit a template? 🤔 This could apply to `TemplateCard`, etc.. Currently there is a mismatch between site and post editors.

## Testing Instructions
Code example:
```js
import { MyTemplateSettingTest } from '@wordpress/edit-site';
const MyTemplateSettingTest = () => (
	<PluginTemplateSettingPanel>
		<p>Hello, World!</p>
	</PluginTemplateSettingPanel>
);


 ```
Render the above component somewhere in your code.

### Example screenshot
Screenshot 2023-05-15 at 2 03 31 PM

### CODE REVIEW

File: packages/edit-site/src/components/sidebar-edit-mode/index.js [hunk 1]
```diff
@@ -74,9 +76,18 @@ export function SidebarComplementaryAreaFills() {
...
-					</PanelBody>
+					<>
+						<PanelBody>
+							<TemplateCard />
+							<PanelRow
+								header={ __( 'Editing history' ) }
+								className="edit-site-template-revisions"
+							>
+								<LastRevision />
+							</PanelRow>
```
Comment (nerrad) [inline packages/edit-site/src/components/sidebar-edit-mode/index.js:87]: This is new to this PR. Did it slip in accidentally from another PR or was it intentional? I didn't see it mentioned in the PR discussion so thought I'd ask :)

Comment (nerrad) [reply packages/edit-site/src/components/sidebar-edit-mode/index.js:87]: Oh nvm, I see it was moved here from `template-card/index.js`.

Comment (ntsekouras) [reply packages/edit-site/src/components/sidebar-edit-mode/index.js:87]: Oh.. I forgot to comment about it, thanks! Yes, I just moved it outside of TemplateCard component because I think it shouldn't be a part of that component.

**Comment by carolinan** [comment]
Always yes to reducing mismatch 😄

File: packages/edit-site/src/components/plugin-template-setting-panel/index.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+PluginTemplateSettingPanel.Slot = Slot;
+
+/**
+ * Renders items in the Template Sidebar below the main information
+ * like the Template Card.
+ *
+ * @example
+ * ```jsx
+ * // Using ESNext syntax
+ * import { MyTemplateSettingTest } from '@wordpress/edit-site';
```
Comment (Mamaduka) [inline packages/edit-site/src/components/plugin-template-setting-panel/index.js]:
```suggestion
 * import { PluginTemplateSettingPanel } from '@wordpress/edit-site';
```

File: packages/edit-site/src/components/plugin-template-setting-panel/index.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/**
+ * Defines an extensibility slot for the Template sidebar.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { createSlotFill } from '@wordpress/components';
+
+export const { Fill, Slot } = createSlotFill( 'PluginTemplateSettingPanel' );
```
Comment (Mamaduka) [inline packages/edit-site/src/components/plugin-template-setting-panel/index.js]:
```suggestion
const { Fill, Slot } = createSlotFill( 'PluginTemplateSettingPanel' );
```
We probably don't need to export these.

**Review by Mamaduka — APPROVED** [review]
> Currently there is a mismatch between site and post editors.

I assume this is **why** we're introducing the new slot :)

The `PluginDocumentSettingPanel` also handles rendering the `PanelBody` wrapper. Do you know if this is something we also want to match here?

Personally, I like devs explicitly handling the rendering of panels, but we might want to mention it somewhere in the docs.

> Should we add this slot in block-editor package in order to be reused when we edit a template?

Maybe not at the moment.

Similar slots are usually contained in the editors. Moving the `PluginDocumentSettingPanel.Slot` into the `block-editor` package and reusing it here might cause features meant for the post editor to leak into the templates.

**Comment by ntsekouras** [comment]
Thanks for reviewing George!

>I assume this is why we're introducing the new slot :)

Yes.

>The PluginDocumentSettingPanel also handles rendering the PanelBody wrapper. Do you know if this is something we also want to match here?

I don't think so. Consumers should be responsible of adding and styling their components accordingly.

**Comment by mcsf** [comment]
Would be good to add a screenshot to this PR for future reference of what part of the UI it concerns. Should it also be marked for dev note?

**Comment by gigitux** [comment]
Hey, @ntsekouras, we are using `PluginTemplateSettingPanel` in WC Blocks to show a button on the Site Editor.  This is the PR https://github.com/woocommerce/woocommerce-blocks/pull/9386, and [this is the logic ](https://github.com/woocommerce/woocommerce-blocks/pull/9386/files#diff-4d5036b9b12b1f89724528a06d64a14ba678df630d32ad1056f08d3aa614a4f0R1)for adding the button on the site editor.

We noticed that importing the `PluginTemplateSettingPanel`, some redux stores are initialized, and this leads to issues that we fixed with a follow-up PR: https://github.com/woocommerce/woocommerce-blocks/pull/9468. 

Even with the fix that I shared above when the `PluginTemplateSettingPanel` is imported and the user visits the post/page editor, this error raises in the console:

```
 Store "core/interface" is already registered.
```

image

I suggest moving the component to another path so that when the developer imports it, no side-effect runs. What do you think?

**Comment by youknowriad** [comment]
@gigitux it seems you might be "bundling" the edit-site dependency which duplicates some JS packages, no?

**Comment by gigitux** [comment]
We use the package `@wordpress/dependency-extraction-webpack-plugin' to extract the @wordpress dependecies. ([source code](https://github.com/woocommerce/woocommerce-blocks/blob/c1486a628527643c096e422b9025209960cac46f/bin/webpack-configs.js/#L54-L61))

I will investigate more, and I will let you know!

**Comment by nefeline** [comment]
Hi there @ntsekouras ! We are still facing issues with the usage of `PluginTemplateSettingPanel` within WooCommerce Blocks, and I was hoping you could help us out with some ideas on how to circumvent these limitations: 

In addition to the error @gigitux [mentioned over here](https://github.com/WordPress/gutenberg/pull/50257#issuecomment-1549279797) that is still ongoing, there's also a conflict with the core's navigation block recently uncovered: whenever you add the Navigation block to any post while having a component importing from `'@wordpress/edit-site';`, which is a requirement for the usage of `PluginTemplateSettingPanel` (example of implementation [over here](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/templates/revert-button/index.tsx#L4)), the following error is triggered:

`edit-site.js?ver=9410ee7f59a6a9b2aa87:3578 Uncaught TypeError: Cannot destructure property 'params' of 'navigation_menu_edit_useLocation(...)' as it is undefined.`

Screenshot 2023-07-24 at 17 49 01

We have been trying to circumvent this in different ways. In order to isolate the changes to the relevant template we tried:

* Conditionally loading `'@wordpress/edit-site'` via PHP;
* Dynamic import to load the module on-demand, at runtime, instead of statically at the beginning.

But no luck so far. Since [we are following](https://make.wordpress.org/core/2023/07/18/miscellaneous-editor-changes-in-wordpress-6-3/) the recommended steps for adding information in the site editor’s Template sidebar, would it make sense to address this limitation on GB?

Thanks in advance!

**Comment by ntsekouras** [comment]
>In addition to the error @gigitux https://github.com/WordPress/gutenberg/pull/50257#issuecomment-1549279797 that is still ongoing, there's also a conflict with the core's navigation block recently uncovered:

Yeah, it seems we have more side effects if we add the `edit-site` package in the post editor, and we could have more similar filters like the one in Navigation block.

I'm not an expert in the bundling tools, but it seems to me that the edit-site shouldn't be bundled when in post editor and secondly maybe displaying block information in this specific slot might not be the best. I'm not sure if we have more extension points in the block inspector sidebar(if we don't, we should), but it feels it should be a better fit there.

The specific slot is for plugins in edit site and it display information under the template card, which would be more fitting for template and site editing related actions.

I'll cc @gziolo and @jsnajdr if they have some input regarding the bundling.

**Comment by jsnajdr** [comment]
When a plugin script imports something from a package, like Woo Blocks' `templates/revert-button` does:
```js
import { PluginTemplateSettingPanel } from '@wordpress/edit-site';
```
Then that package becomes the plugin script's dependency. Not only the package imports are turned into externalized references like `wp.editSite.PluginTemplateSettingPanel`, which is desired, but also the `edit-site` script is enqueued and executed. And that's undesired -- in an app like `edit-post` or `edit-widgets`, we don't want to load _another_ app like `edit-site`. They will conflict with each other, there is a lot of static initialization code that registers data stores etc.

Because `edit-site` or `edit-post` are not merely library packages that export some useful slots for plugins, they are apps.

The same problem will arise when a plugin is loaded into the `edit-site` environment and will reference some slot (like `PluginPostPublishPanel`) from `edit-post`. Here again, the Post Editor app will be loaded into the Site Editor app.

A good solution for this is that the plugin makes sure that scripts that reference the `edit-site` package are enqueued only in the Site Editor and scripts that reference `edit-post` are enqueued only in Post Editor.

Woo Blocks doesn't do this -- the offending `edit-site` import is happening inside the `assets/js/index.js` module, and that module is being bundled into a shared `wc-blocks` script ([here](https://github.com/woocommerce/woocommerce-blocks/blob/a9a9e648172e9f5225543fb0e842b96886ed3ec5/bin/webpack-entries.js#L146)) which is a dependency of just about everything else.

Instead, Woo Blocks should build `templates/revert-button` as a separate script. It's an independent standalone plugin. And then enqueue it only in the Site Editor, nowhere else. That should happen in PHP code, somewhere in `AssetsController.php`.

**Comment by nefeline** [comment]
Thank you so much @jsnajdr and @ntsekouras for your prompt feedback and for sharing all these details! Inspired by your input, the following solution  was proposed for this problem: https://github.com/woocommerce/woocommerce-blocks/pull/10388

**Comment by youknowriad** [comment]
I just stumbled on this Slot, I think this should have been just the `PluginDocumentSettingPanel` but expanded to the template sidebar (both in post and site editors). In other words, plugin authors already discriminate using postType in `PluginDocumentSettingPanel` if they want to do post type specific panels. It should have been the same for this one.

That said, there's potentially a path forward where the implementation of `PluginTemplateSettingPanel` internally is just `PluginDocumentSettingPanel` + an additional condition on post type.

**Comment by ntsekouras** [comment]
>That said, there's potentially a path forward where the implementation of PluginTemplateSettingPanel internally is just PluginDocumentSettingPanel + an additional condition on post type.

That's a good thought, although I'm not really sure how would we proceed and keep backwards compatibility, since [PluginDocumentSettingPanel](https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-document-setting-panel/) ~relies on `registerPlugin`. Currently `PluginTemplateSettingPanel` consumers just use the slot without any additional APIs~.



Besides the above, I observed now that at some point where we introduced the `PagePanels` and `TemplatePanel` we register this slot for both, which should only be rendered for templates. I'll open a PR for that for now..

**Comment by Mamaduka** [comment]
Using `registerPlugin` isn't a hard requirement for the `PluginDocumentSettingPanel` SlotFill. It's just a registration method we recommend for third-party consumers.

**Comment by ntsekouras** [comment]
>Using registerPlugin isn't a hard requirement for the PluginDocumentSettingPanel SlotFill. It's just a registration method we recommend for third-party consumers.

Hm.. I thought I tried that in code.. Let me try again 😄

**Comment by ntsekouras** [comment]
As @Mamaduka pointed out registerPlugin is optional, but the issue here for me is:
`PluginDocumentSettingPanel` checks if the panel is opened, which is `false` by default and that results in **not** showing the contents. The other difference is that this slot didn't render the `PanelBody` whereas PluginDocumentSettingPanel does. 

It's not clear for me what would be the best path forward. I'll have to think more about it, but any suggestions are welcome.

**Comment by youknowriad** [comment]
Maybe the best path is just to soft deprecate this one.
